### ISSUE TITLE
Decoding big panorama JPGs is broken

### ISSUE THREAD

**Issue by derselbst** [t0rakka/mango#73]
I tried to run the `jpeg_benchmark` on a JPG with 63939 x 10705 pixels and was greeted with a `std::bad_alloc`. Investigation showed that this is coming from an arithmetic overflow:

```gdb
(gdb) bt
#0  0x00007ffff741763e in __cxa_throw () from /usr/lib64/libstdc++.so.6
#1  0x00007ffff740b7c6 in ?? () from /usr/lib64/libstdc++.so.6
#2  0x00000000004aa02e in (anonymous namespace)::load_surface (surface=..., memory=..., extension=".jpg", format=0x0) at mango/source/mango/image/surface.cpp:192
#3  0x00000000004aa13f in (anonymous namespace)::load_surface (surface=..., filename="AAAD2760.dpp - AAAD2902.dpp_fused_enblend.jpg", format=0x0)
    at mango/source/mango/image/surface.cpp:203
#4  0x00000000004ab5d1 in mango::Bitmap::Bitmap (this=0x7fffffffd5d0, filename="AAAD2760.dpp - AAAD2902.dpp_fused_enblend.jpg") at mango/source/mango/image/surface.cpp:569
#5  0x0000000000405df8 in main (argc=<optimized out>, argv=0x7fffffffdb58) at mango/example/image/jpeg_benchmark/jpeg_benchmark.cpp:259
(gdb) frame 2
#2  0x00000000004aa02e in (anonymous namespace)::load_surface (surface=..., memory=..., extension=".jpg", format=0x0) at mango/source/mango/image/surface.cpp:192
192                 surface.image  = new u8[surface.height * surface.stride];
(gdb) print surface
$1 = (mango::Surface &) @0x7fffffffd5d0: {format = {bits = 32, type = mango::Format::UNORM, flags = 0, size = {component = "\b\b\b\b", {r = 8 '\b', g = 8 '\b', b = 8 '\b', a = 8 '\b'}}, offset = {component = "\020\b\000\030", {
        r = 16 '\020', g = 8 '\b', b = 0 '\000', a = 24 '\030'}}}, image = 0x0, stride = 255756, width = 63939, height = 10705}
(gdb) print surface.height * surface.stride
$2 = -1557099316
```

Working it around by casting to `size_t` showed promising results:

```shell
----------------------------------------------
                load         save             
----------------------------------------------
libjpeg:   10275.5 ms    4455.0 ms 
stb:       12761.3 ms   34738.5 ms 
jpgd:          0.6 ms     154.5 ms 
mango:      7585.8 ms    2092.4 ms
```

Unfortunately, output-mango.jpg has a missing black part at the bottom of the image:

![Screenshot_20201024_133545](https://user-images.githubusercontent.com/8152480/97080762-11d54c80-15fe-11eb-8ff3-51414ce0ba5f.png)

I cannot provide the panorama input image. However, any other artifically generated image with a similar size should show the same problems.

**Comment by t0rakka** [comment]
It does come from overflow indeed, the image is 2.6 GB which is larger than largest presentable signed 32 bit value: it won't fit! I rather keep the width and height 32 bit integers so the storage amount should be size_t as that is the type used for that kind of stuff. This will of course keep failing on 32 bit architectures when the requested storage is over 4 GB but they literally can't do more so that's fine by me.

At least it should work when it can work.. so.. I'll patch the storage types to be size_t where needed. Let's see if we can get this working..

**Comment by t0rakka** [comment]
OK, generated a large image, saved as jpeg, but macOS can't display it.. copying it into Linux or Windows to see what it looks like and fix any errors (I am testing and fixing the saving first).

After save + load works, I'll also go through the functions that allocate storage in surface.cpp since some of them use width and height to compute the storage and also need to be updated.

**Comment by t0rakka** [comment]
Saving tested: works OK.

**Comment by t0rakka** [comment]
Saving tested: works OK.
Disclaimer: tested on 64 bit macOS, 

uname -a

Darwin MacBook-Pro-3 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64

"int" is signed 32 bit so it *should* overflow, but it doesn't .. strange..

What OS and compiler did you use for reproducing this?

**Comment by t0rakka** [comment]
Meanwhile, I test on Linux on 64 bit Intel and 32 bit ARM (I happen to have ARM devboard around I can compile as 32 bit, I'd rather not install 32 bit VM hosted OS if I can avoid it).

Yeah, running a bit short on 32 bit computers these days.. :) 

As last resort I can install 32 bit OS on a VM.

**Comment by t0rakka** [comment]
If you can't send the image, can you enable debug prints for JPEG reader from jpeg.hpp, recompile library and send me the loader debug output. It might be a progressive image or something else like that so I could focus my search on the right areas.

**Comment by derselbst** [comment]
```shell
uname -a
Linux main 5.3.18-lp152.44-default #1 SMP Wed Sep 30 18:51:43 UTC 2020 (914f31e) x86_64 x86_64 x86_64 GNU/Linux
```

openSUSE Leap 15.2, an LP64 machine, GCC 7.5.0, using `cmake -DENABLE_AVX=1`

You're right, it's a progressive one indeed, containing an embedded exif thumbnail preview. The debug log: [debug.log](https://github.com/t0rakka/mango/files/5433202/debug.log)

**Comment by t0rakka** [comment]
I committed and pushed some fixes but since I can't reproduce let me know if they help. I think the progressive is still broken so the end of the buffer should be broken, looking into it next. Meanwhile it would be good to hear if the changes I made fix the original issue.

I changed the stride computation to use size_t intermediately, but I won't store stride as size_t as it is unsigned type; stride has to be signed so that it can go both up and down in the image. I know unsigned stride could also go up, have same effect as negative signed stride because of wrap-around but I prefer to keep it signed for clarity. :)

**Comment by t0rakka** [comment]
Yeah, I was suspecting the missing bottom has to be something less common as the load+save worked for me. :D

**Comment by t0rakka** [comment]
Your image width is super close to limits, btw.

        if (xsize <= 0 || ysize <= 0 || xsize > 65535 || ysize > 65535)
        {
            // NOTE: ysize of 0 is allowed in the specs but we won't
            header.setError("Incorrect dimensions (%d x %d)", xsize, ysize);
            return;
        }

JPEG specification limits maximum image dimensions to 65535. I added a lot of checks for incorrect files for Fuzz testing so that incorrect files don't crash or do other nasty stuff. The code is not 100% proof but it stops 99.99% of incorrect files cold. The Huffman decoder also takes a small performance hit because of this but I felt the tradeoff was worth it as the effect wasn't pronounced.

**Comment by t0rakka** [comment]
The image is really large, almost 3 GB.. and the progressive decoder is the worst case as it needs to keep a temporary copy of the full image during loading (as it is progressively filling the frequencies into the buffer). The frequency data is 16 bits per channel (of Y, Cb, Cr) so it's 6 GB worth of temp buffers.. outch.. and there is no way around that either due to the way the format is designed. :( 

Sorry about that.. I recommend storing the image as non-progressive to save memory at loading time, if that is ever an issue. The non-progressive loader writes directly into the target surface and tries to minimize memory usage (however, if you are highly threaded each thread has its own scratch buffer).

I think the memory usage can be reduced a bit better but that's for some other time. I'm also looking for alternatives for work scheduling so that threading overhead can be reduced that is not very obvious source of pain for 4-16 core computers but recently I got AMD Threadripper 3990x with 128 threads and the overhead begins to hurt at around 48 threads or so.. a solution is sorely needed to scale to higher thread counts. But that is also something else, just letting you know such sore spots are well known now. :)

**Comment by t0rakka** [comment]
If I reduce the thread pool to 32 threads, the overhead is negligible -but- then I need 4 of them and distributing work evenly to four pools is really hard. If I just split work into 4 it's often that 1 pool keeps going long after others are done. The communication between them is the problem: threads communicate with each other by writing into memory, and any write has to be propagated to caches of other cores when they read or write into the same location. 128 threads doing this is what kills the performance, so the design has to spread work evenly w/o knowing what the other threads are doing. See what the problem is here? =)

Only way for 128 threads to work efficiently is that they work in complete isolation from each other, and that isolation makes it super hard to distribute work evenly.. interesting problem to crack so no hard feelings, AMD. :)

**Comment by derselbst** [comment]
Ok, I checked out recent master and I'm sorry to say but it seems you've broke it :)

```gdb
image: 63939 x 10705 (95226 KB)
----------------------------------------------
                load         save             
----------------------------------------------

libjpeg:   10334.8 ms    3782.6 ms 
stb:       12559.6 ms   34643.8 ms 
jpgd:          0.0 ms       0.0 ms 

Thread 5 "jpeg_benchmark" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff525b700 (LWP 7873)]
0x0000000000508d1f in std::__atomic_base<unsigned long>::load (__m=std::memory_order_relaxed, this=0xfffffefaffffff12) at /usr/include/c++/7/bits/atomic_base.h:396
396             return __atomic_load_n(&_M_i, __m);
Missing separate debuginfos, use: zypper install libgcc_s1-debuginfo-9.3.1+git1296-lp152.2.3.1.x86_64 libjpeg8-debuginfo-8.1.2-lp152.8.3.1.x86_64 libstdc++6-debuginfo-9.3.1+git1296-lp152.2.3.1.x86_64
(gdb) y
Undefined command: "y".  Try "help".
(gdb) bt
#0  0x0000000000508d1f in std::__atomic_base<unsigned long>::load (__m=std::memory_order_relaxed, this=0xfffffefaffffff12) at /usr/include/c++/7/bits/atomic_base.h:396
#1  moodycamel::ConcurrentQueue<mango::ThreadPool::Task, moodycamel::ConcurrentQueueDefaultTraits>::ProducerBase::size_approx (this=0xfffffefafffffef2)
    at /home/tom/Programme/mango/source/mango/core/../../external/concurrentqueue/concurrentqueue.h:1719
#2  0x0000000000507209 in moodycamel::ConcurrentQueue<mango::ThreadPool::Task, moodycamel::ConcurrentQueueDefaultTraits>::try_dequeue<mango::ThreadPool::Task> (this=0x9f5488, item=...)
    at /home/tom/Programme/mango/source/mango/core/../../external/concurrentqueue/concurrentqueue.h:1112
#3  0x00000000005028cd in mango::ThreadPool::dequeue_and_process (this=0x9d9fc0 <mango::ThreadPool::m_static_instance>) at /home/tom/Programme/mango/source/mango/core/thread.cpp:235
#4  0x000000000050265a in mango::ThreadPool::thread (this=0x9d9fc0 <mango::ThreadPool::m_static_instance>, threadID=3) at /home/tom/Programme/mango/source/mango/core/thread.cpp:187
#5  0x00000000005021ec in mango::ThreadPool::<lambda()>::operator()(void) const (__closure=0x9fbf38) at /home/tom/Programme/mango/source/mango/core/thread.cpp:128
#6  0x0000000000503f7b in std::__invoke_impl<void, mango::ThreadPool::ThreadPool(size_t)::<lambda()> >(std::__invoke_other, mango::ThreadPool::<lambda()> &&) (__f=...) at /usr/include/c++/7/bits/invoke.h:60
#7  0x0000000000503a2b in std::__invoke<mango::ThreadPool::ThreadPool(size_t)::<lambda()> >(mango::ThreadPool::<lambda()> &&) (__fn=...) at /usr/include/c++/7/bits/invoke.h:95
#8  0x0000000000504c5c in std::thread::_Invoker<std::tuple<mango::ThreadPool::ThreadPool(size_t)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x9fbf38) at /usr/include/c++/7/thread:234
#9  0x0000000000504ba9 in std::thread::_Invoker<std::tuple<mango::ThreadPool::ThreadPool(size_t)::<lambda()> > >::operator()(void) (this=0x9fbf38) at /usr/include/c++/7/thread:243
#10 0x0000000000504b1c in std::thread::_State_impl<std::thread::_Invoker<std::tuple<mango::ThreadPool::ThreadPool(size_t)::<lambda()> > > >::_M_run(void) (this=0x9fbf30) at /usr/include/c++/7/thread:186
#11 0x00007ffff7441140 in ?? () from /usr/lib64/libstdc++.so.6
#12 0x00007ffff7bc04f9 in start_thread () from /lib64/libpthread.so.0
#13 0x00007ffff6b5afbf in clone () from /lib64/libc.so.6
(gdb) frame 1
#1  moodycamel::ConcurrentQueue<mango::ThreadPool::Task, moodycamel::ConcurrentQueueDefaultTraits>::ProducerBase::size_approx (this=0xfffffefafffffef2)
    at /home/tom/Programme/mango/source/mango/core/../../external/concurrentqueue/concurrentqueue.h:1719
1719                            auto tail = tailIndex.load(std::memory_order_relaxed);
(gdb) list
1714
1715                    inline ProducerBase* next_prod() const { return static_cast<ProducerBase*>(next); }
1716
1717                    inline size_t size_approx() const
1718                    {
1719                            auto tail = tailIndex.load(std::memory_order_relaxed);
1720                            auto head = headIndex.load(std::memory_order_relaxed);
1721                            return details::circular_less_than(head, tail) ? static_cast<size_t>(tail - head) : 0;
1722                    }
1723
(gdb) print tailIndex 
Cannot access memory at address 0xfffffefaffffff12
(gdb) print headIndex
Cannot access memory at address 0xfffffefaffffff1a
(gdb) info threads
  Id   Target Id                                         Frame 
  1    Thread 0x7ffff7fbb740 (LWP 7866) "jpeg_benchmark" 0x00000000004c55ca in _mm_srai_epi32(long long __vector(2), int) (__B=17, __A=...) at /usr/lib64/gcc/x86_64-suse-linux/7/include/emmintrin.h:1168
  2    Thread 0x7ffff6a5e700 (LWP 7870) "jpeg_benchmark" _mm_sub_epi16(long long __vector(2), long long __vector(2)) (__B=..., __A=...) at /usr/lib64/gcc/x86_64-suse-linux/7/include/emmintrin.h:1072
  3    Thread 0x7ffff625d700 (LWP 7871) "jpeg_benchmark" 0x0000000000508d1f in std::__atomic_base<unsigned long>::load (__m=std::memory_order_relaxed, this=0xff90ffffff910017) at /usr/include/c++/7/bits/atomic_base.h:396
  4    Thread 0x7ffff5a5c700 (LWP 7872) "jpeg_benchmark" mango::jpeg::idct_sse2 (dest=0x7ffff5a5b7c0 '\377' <repeats 64 times>, src=0x7fff19536ec0, qt=0xa03340) at /home/tom/Programme/mango/source/mango/jpeg/jpeg_idct.cpp:304
* 5    Thread 0x7ffff525b700 (LWP 7873) "jpeg_benchmark" 0x0000000000508d1f in std::__atomic_base<unsigned long>::load (__m=std::memory_order_relaxed, this=0xfffffefaffffff12) at /usr/include/c++/7/bits/atomic_base.h:396
```

I wanted to provide a core dump, but it's 4GiB, so forget about it. I have to leave for today anyway, will be back tomorrow.

P.S.: For me, memory is not a limitation. I have 24 GiB of it. I prefer progressive (esp. for big images) because they support incremental display during the decoding process. Unfortunately, I haven't found an image viewer that actually supports that yet.

**Comment by t0rakka** [comment]
Noticed.. :P

**Comment by t0rakka** [comment]
Noticed; fixed now - use values before initialisation.

I did two things to the progressive decoder:

1. The temp buffer allocation now supports larger buffers
2. The progressive post-process step allows larger offsets for the MCUs now

These should improve the situation but as I don't have such large progressive file..

**Comment by t0rakka** [comment]
image: 63939 x 10705 (165128 KB)
----------------------------------------------
                load         save             
----------------------------------------------
libjpeg:    6658.6 ms    8066.6 ms 
stb:        5915.4 ms   25930.7 ms 
jpgd:          0.8 ms       0.5 ms 
mango:       761.3 ms     695.1 ms

**Comment by t0rakka** [comment]
```
image: 63939 x 10705 (165128 KB)
----------------------------------------------
                load         save             
----------------------------------------------
libjpeg:    6658.6 ms    8066.6 ms 
stb:        5915.4 ms   25930.7 ms 
jpgd:          0.8 ms       0.5 ms 
mango:       761.3 ms     695.1 ms 
```

**Comment by t0rakka** [comment]
Also fixed the print in example to use zu/size_t for the size in bytes..  two things worth taking note of:

1. The progressive has to work; let me know how it is doing
2. When the image has RST markers in it, loading is MUCH faster (see above results)

RST marker is basically a ReSTart interval marker in the JPEG file which restarts the Huffman decoder. The loader will detect these and fire off a Huffman decoder thread to handle these decoded segments individually so when you have more CPU cores they each can decode their own part of the file independently. In above timings I am on 6 core macbook pro which can do 12 threads and we get ~9x speed up, it's really sweet. So if you're looking for speed the files *should* have RST markers in them. The MANGOs jpeg encoder always uses RSTs to accelerate the encoding (up there the speed up is 11.6x which is really close to theoretical max speedup on this hardware!)

Use of RSTs is highly recommended when you can control the files you use.

**Comment by t0rakka** [comment]
Tested progressive 60K x 10K image loading: OK with the latest changes; still need to hear how it works with your specific image.

**Comment by derselbst** [comment]
Ok, the segfault is fixed. However, the black part has grown :(

I tried your RST marker idea. And yes, you are much faster, but in my case, mango's output image is entirely black. Since this seems to be hard to solve without the input, I'm providing you the with two test images. (Censored and in the worst quality possible, to avoid stealing it.)

[ITs-A-TAR-FILE!.zip](https://github.com/t0rakka/mango/files/5434377/ITs-A-TAR-FILE.zip)

**Comment by t0rakka** [comment]
I'll delete the files after I'm done checking this out.

**Comment by t0rakka** [comment]
Both of the provided files now load correctly on my test bench as far as I can tell. The problem was seekMarker which should skip RSTs as they are part of the compressed stream.

The decoder has a look-ahead feature for compressed stream handling so it can be ahead by up to 6 bytes. I corrected for this rewinding 6 bytes and seeking for the next marker so that parsing can continue. Unfortunately the files you provided had unlucky RSTs just at the end of the compressed segment; super weird since there should be no valid reason for any RSTs be there - but there was - and the code is now more resilient.

The RST interval was 65535 which is a really strange interval, usually it is number of MCU*blocks_per_mcu for each scan. This is really strange number.. but it looks like this file indeed has RST every 65535 blocks and it is a valid file.

Another thing this prompted me to look into: the progressive decoder doesn't benefit from the RSTs, never thought this would be useful scenario to optimise for since it's a bit tricky one but I think it's manageable; each SOS (Start of Scan) has implicit restart which should keep things simple enough.. TL;DR - I take a look if I can give this path a speed boost but won't happen right away.

I look forward hearing that the code works for you now.. if not.. we're back to debugging mode.. but I confirm that I could reproduce the issue with the files you provided and corrected the problems I found.

**Comment by derselbst** [comment]
Ok, the good news: The `pano-prog-rst.jpg` doesn't result in a black image anymore. The bad news: For both images the black missing bottom part is still there. To make it worse, the `pano-prog-rst.jpg` now has a few pixel rows of random noise in the middle of the picture. This one is hard to reproduce, though. Here's the `output-mango.jpg` obtained by jpeg_benchmark, so you see what I mean (too big for github, thus gdrive):

https://drive.google.com/file/d/1DO3fiQipH13ER4nzDUIsgNjJDRHOf56k/view?usp=sharing

I wanted to be sure, so I deleted all mango files and recompiled with `cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_AVX=1 ..`, but the problems persists.

> The RST interval was 65535 which is a really strange interval, usually it is number of MCU*blocks_per_mcu for each scan.

FYI, I obtained the image by exporting it from GIMP. They provide an option called "Intervall (MCU-rows)" which is set to a default of 16. I kept this default.

**Comment by t0rakka** [comment]
I created unstable branch "jpeg_panorama" where I am rewriting the parser a little bit. I modified the getBits / peekBits to ensure-before-use and removed all of the built-in ensure's from the Huffman decoder.

The fill() which fills the register is already terminating pointer advancing at marker, so the pointer shouldn't "overflow", so the seekMarker() is not needed anymore ; deprecated.

The code detects "stuff bytes" correctly so they shouldn't cause any problems. RST marker will terminate block processing but that shouldn't be a problem either as there shouldn't be RSTs mid-block. They are literally used for MCU interval restart so getting one while decoding bitstream would be incorrect bitstream anyway.

Feel free to test with the branch head and let me know which commit work, if it does. It's a bit rough to fix this w/o test vectors so I'll do an audit on the whole parser.. cleanup on the code is long overdue anyway..

So if the head starts to work by some miracle let me know.. the work is just starting and haven't ran the validations yet so that no idea about regressions yet.. I'm doing that on a different desktop machine.. it's not automatic like Jenkins validation we use at work, this is just a side project..

**Comment by t0rakka** [comment]
The image has 7993 MCUs per scan (scan is 8 pixels high, since MCU size is 8x8). 16 scans per restart interval yields this:

7993 * 16 = 127888

The file reports 65535, which is 0xffff ; largest presentable unsigned 16 bit value. I think the encoder clamps the interval to this.. so the RST markers are in middle of scans. This is completely legal and the code didn't assume that they would always be at start of scan anyway.. just clarifying where the "mystery" value comes from.

I looked at optimizing the progressive decoding loop and it's not that difficult when took a good look at it. The main thing to do is to linearize the calculations in the loop. Currently we have outer y-loop and inner x-loop, and values are computed for the x-loop.

Since the RSTs can be detected almost anywhere, the code has to be structured so that it does only one loop for ALL MCUs so that we can split the range into digestible chunks which then can be queued for parallel execution. 

Next steps:

1. optimize the progressive decoder loop
2. setup more tests for the regression testing
3. verify that the large panorama images work
    FAIL -> more fixing in horizon
    PASS -> \ :D /

**Comment by t0rakka** [comment]
First version of parallel Progressive-RST decoder committed to the "jpeg_panorama" branch. There is a lot of duplication in the code so clean up and refactoring is still coming soon, but as proof-of-concept it does the trick.

If the decoder still fails the debug output logs would be cool. Also enable the prints at end of processSOS, which print out 16 bytes around the SOS segment end. There is #if 0 which needs to be toggled on for this info to be printed. I can see if we hit or miss markers with the new parsing system.

The root cause for the missing pixels seems to be that the parser goes off-sync and exits early.. so the prints will tell me if that is still happening and what's around the parsing position when it happens.

**Comment by t0rakka** [comment]
This would be so much faster in a debugger, but no problem, you did find a bug and I want to fix it so your help is appreciated, thanks already for the assistance so far!!! <3

**Comment by derselbst** [comment]
Thanks for investigating this.

I've tried the jpeg_panorama branch. Unfortunately, it doesn't improve the situation. The missing bottom part is still there for both images. And I also managed to reproduce the garbage pixels for the `pano-prog.jpg` now. So, here's the new picture with garbage pixels + debug log from the decoder.

[pano-prog-with-garbage-pixels.log](https://github.com/t0rakka/mango/files/5439589/pano-prog-with-garbage-pixels.log)
[output-mango-pano-prog-new.zip](https://github.com/t0rakka/mango/files/5439594/output-mango-pano-prog.zip)

For completeness I'm also attaching the debug log of the `pano-prog-rst.jpg`, which this time only had the missing bottom part after decoding:

[pano-prog-rst.log](https://github.com/t0rakka/mango/files/5439611/pano-prog-rst.log)

**Comment by t0rakka** [comment]
Let's take a peek at one of the SOS segment ends in the RST log:

ff 00 d7 ff d2 a7 e3 7f | ff c4 00 33 10 00 02 01

ff 00   <-- a "stuff byte", this is used to write a literal value 0xff into the stream.. 0xff is a marker in jpeg stream always followed by identifier. The only case when it is a literal value is when it is followed by a 0, which marks it as a value instead of marker (the zero is skipped).

ff d2 <-- restart marker #2 (they are numbered from 0 to 7, and repeat as 3 lowest bits of the interval index.. this is how the decoder is supposed to know how many MCUs to skip when there is corruption in the stream)

| <-- marks where the parsing pointer is at after parsing SOS segment, as we can see there is "ff ??" (a marker) after every SOS segment in the file so the parser seems to be working OK

The last clue is that the EOS (ff d9) is correctly detected at the end, so we haven't missed any SOS segments. The seekMarker() is now deprecated so we don't accidentally skip anything we shouldn't. The only place where we can screw this up now is seekRestartMarker, have to go through that function super carefully. :) :) 

ff 00 d7 ff d2 a7 e3 7f | ff c4 00 33 10 00 02 01

Quite interesting there are only 3 bytes in the last interval here: a7 e3 7f (the ff d2 was a restart marker), worth looking into in more detail..

fe ff 00 fe ff d2 1f 8d | ff c4 00 15 11 01 01 00 

ff c4 <-- huffman table declaration

Some of the SOS segments have huffman table after them, which means a different huffman table should be used.. I might have to move the table into the DecodeState for the multi-threaded case since we now support parallel decoding and can't modify the table while it's in use by someone else..  

Here the table changes in middle of restart interval.. no problem, I just need to update the table in current decoder state when it changes, and any time I start a new interval copy it for the worker thread. If the worker encounters a new table declaration it needs to make it active for any thread that is going to work after that point in the file.. so.. I need to threat the huffman table marker (ff c4) as restart marker so that at that point the state is set active and copied to any worker than starts after that point.

OK, yeah, gotcha.. so Huffman Table markers have to also begin a new interval.. and that should do the trick..

FYI, most of the rambling here is about the RST variant.. I get that we also have problem in the serial non-RST variant so there is still some glitch. but fixing the parallel code ASAP. :)

Parsing itself seems fine based on what I can see on the logs, unless some interval is completely skipped by accident (seekRestartMarker needs careful looking into as mentioned earlier).

SO HARD to debug like this, but on the bright side the code is getting a fairly comprehensive audit + upgrade as a side-effect, so that's nice. :)

**Comment by t0rakka** [comment]
I'll try to create a test file with GIMP with same dimensions.. should have done that yesterday..

**Comment by jukka-liimatta** [comment]
GIMP did it; I can reproduce nicely. First thing I did was clear the surface to blue before decoding, and behold, blue bar at the bottom.. but not before crashing!

The clear() member function crashes; it computes the address like this:

scan = image + y * stride

This, as we know, wraps around the 32 bit signed integer offset.. so bottom of image was never touched, and also explains corruption.

Since this will be ENDLESS swamp of casting to size_t everywhere, I just surrender and use size_t for stride in mango::Surface. This way the fix propagates everywhere, and if it causes compiler warnings they can be corrected as compiler will now spot them for us.

I hope to have something soon.. meanwhile, all it took was some crazy dude with his crazy sized images.. so THANK YOU :D

**Comment by jukka-liimatta** [comment]
Fixed the strides everywhere. I get correct picture with the GIMP files I wrote.. :_D

**Comment by jukka-liimatta** [comment]
I disabled the RST-parallel progressive decoding until I tested it more.

It shouldn't have worked if Huffman tables are modified while someone else is still using them. I guess the new uploaded tables are identical but that's just a fluke. I refactor the code to always make a copy of the tables for any thread that uses them (the tables are fairly small, no problem to copy them).

The speedup is promising: clearing the 2.7 GB image takes 750 ms and.. decoding the image takes.. 650 ms... decoding is faster than std::fill_n / memset on the target surface! So yeah, I want that code active ASAP! =P

**Comment by jukka-liimatta** [comment]
Also, this is a good remiinder that even if std::memset is super optimized it doesn't use multichannel memory since it still writes sequentially. Say, if I have a dual channel it would double the write bandwidth to write from two threads into two different regions of memory. Of course, this consumes the bandwidth from other processes and threads.

But it's not always smart strategy to hog every resource for oneself, which is where my scheduler rewrite is going (slow progress).

**Comment by t0rakka** [comment]
Parallel progressive decoder enabled; Huffman tables can't be changed in middle of SOS segment.

**Comment by derselbst** [comment]
I've just tested the recent branch @ b9100ed03527a4d57e922ea89e303a177b26534d and it's working perfectly for me. Also, the speedup that you achieve with RST markers is tremendous! Thanks guys!

**Comment by t0rakka** [comment]
No, thank you! I was completely oblivious about the stride problem with large images.. that was THE problem all the time, but I also got to improve and clean up the JPEG code.

Added a TODO note to check stride usage in the library everywhere. The non-progressive code still have some suspicious stride usage but I need test files to catch them, so will generate some but for now we can close this specific issue.
